### DISCUSSION TITLE
Pi Pico flash upload

### DISCUSSION THREAD

**Discussion by graemecoard** [RudolphRiedel/FT800-FT813#88] [Q&A]
Hi Rudolph, first of all thanks for making this library, I came from using gameduino and it's great to have a library that's active and has frequent updates.

Now I am aware the flash upload part does not relate to your library however you seem to be a guru and the BRT forums are a pain.

I am trying to upload a flash file to a Riverdi RVT70HSBNWN00 using a pi pico. I am encountering multiple issues and I believe they are stemming from the Pico not uploading the flash files correctly.
I am trying to display a full page image 1024x600 in ASTC4X4 format. 
I have encountered a two different problems:
- EAB will show a successful upload but the light on the Pico will still be flashing, and nothing gets uploaded - post a cmd_flasherase, it shows a white empty image, or without a cmd_flasherase, the image is unchanged.
- EAB will show a successful upload but the end of the image is cut off by a few ASTC lines of white. If I pad the file by adding another image on the end, this allows me to show the image.

I have read a few other similar discussions but I think my problem is different.
To display the image, I have had some success using 
```
EVE_init_flash();  
EVE_cmd_flashread(0, 4096, 614400);  
EVE_execute_cmd();
```
In my TFT_init and:
```
EVE_cmd_setbitmap_burst(0, EVE_ASTC_4X4, 1024, 600);   
EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);  
EVE_cmd_dl_burst(VERTEX2F(0, 0));  
EVE_cmd_dl_burst(DL_END);
```

In my burst code within TFT_display.
I have also tried the example of displaying directly from flash memory using your example from #83, and have tried different sizes and formats.

Any help would be appreciated, thanks.

**Comment by graemecoard** [comment]
Also, to further test and find out where the problem is, is it possible include the .bin as a .c as converted by the EAB, then copy it to the flash in a separate program? I see you can do this with CMD_INFLATE but I'd rather not have to compress then convert again, it would be easier just to upload the bytes in PROGMEM and transfer to flash, even if it is 1024 bytes at a time due to the size of RAM_G.
Thanks

**Comment by RudolphRiedel** [comment]
Hello,

I was about to try flashing with PiPico and got interrupted by other things. :-)

This is an issue with either the eve_flash_pico.uf2 or with uf2conv4eve.exe that is supplied with EAB.
The flashing itself is done with CMD_FLASHUPDATE which only works with 4096 byte blocks.

Now eve_flash_pico.uf2 does not seem to be padding to 4096 bytes but uf2conv4eve also is not providing the .uf2 files with padding to 4096 bytes.
For bin files below 8192 bytes the UF2 file is even empty.

So a workaround would be to add a dummy file to the flash image in EAB, 4096 bytes of 0xff: 
[dummy.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11584418/dummy.zip)

I am trying to fix uf2conv4eve.py for a pull-request, this is found here: https://github.com/Bridgetek/pico-brteve
And it probably is the source from which uf2conv4eve.exe is compiled.

>Also, to further test and find out where the problem is, is it possible include the .bin as a .c as converted by the EAB,
>then copy it to the flash in a separate program?
>I see you can do this with CMD_INFLATE but I'd rather not have to compress then convert again,
>it would be easier just to upload the bytes in PROGMEM and transfer to flash,
>even if it is 1024 bytes at a time due to the size of RAM_G.

While it is possible to not use CMD_INFLATE in between by converting the image directly to a .h and using
EVE_memWrite_flash_buffer() to transfer the array from the controller to RAM_G, this does not change anything
in regards of flashing the data to the external chip on the module.
This only changes the amount of data that needs to be stored in the microcontroller.

My example code has this in tft.c to flash the font image:

`
uint32_t datasize;

EVE_cmd_inflate(0, flash, sizeof(flash)); /* de-compress flash-image to RAM_G */
datasize = EVE_cmd_getptr(); /* we unpacked to RAM_G address 0x0000, so the first address after the unpacked data also is the size */
EVE_cmd_flashupdate(0,0,4096); /* write blob first */
if (E_OK == EVE_init_flash())
{
   EVE_cmd_flashupdate(0,0,(datasize|4095)+1); /* size must be a multiple of 4096, so set the lower 12 bits and add 1 */
}
`

So whatever the size of the image file is after de-compressing it to RAM_G, the last line is the most important one here as it is padding the file to 4096.

**Reply by RudolphRiedel** [reply]
Ah, EAB is padding the .bin file but only to 256 bytes, this could also be seen as the issue.

**Comment by RudolphRiedel** [comment]
Try this:
[uf2conv4eve.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11584680/uf2conv4eve.zip)

It pads the .bin file to 4096 and also fills in the padded space with 0xff.

**Comment by RudolphRiedel** [comment]
Hmm, now that the .uf2 is generated correctly I wanted to actually try out flashing with the Pi-Pico.
But feeding the Pi-Pico with eve_flash_pico.uf2 does not result in the BTEVE-FLASH drive to show up.

**Comment by RudolphRiedel** [comment]
Found my problem, in order for eve_flash_pico.uf2 to work it is necessary to actually connect a display.

**Comment by RudolphRiedel** [comment]
For some reason uf2conv4eve.py and uf2conv4eve.exe do not write the b817.blob to an empty flash.
I reported this here: https://github.com/Bridgetek/pico-brteve/issues/28

Using \EVE Asset Builder\flash_blob\bt815.blob does work though.

**Comment by graemecoard** [comment]
Hi, thanks for all your help, unfortunately I cannot get uf2conv4eve.py to work, I'm not sure how to use it correctly? I have tried running it from cmd with arguments and have tried converting to .exe using auto-py-to-exe and placing it in the /tools directory of EAB.

EDIT:
Placed uf2conv4eve.py and my flash.bin in EAB/tools folder.
Copied bt815.blob and bt817.blob from /flash_bob to /tools
Tried a couple of commands as in your issue in [https://github.com/Bridgetek/pico-brteve/issues/28](url) after erasing flash:
python uf2conv4eve.py -f BT815 bt815.blob -i flash.bin
-> this appears to work, uploads, but still can't load the file and EVE_init_flash() returns 8 (EVE_FAIL_FLASHFAST_NO_HEADER_DETECTED)
python uf2conv4eve.py -f BT817 bt817.blob -i flash.bin
-> appears to upload but light on pico keeps flashing and BTEVE-FLASH does not go away

Just to confirm, I loaded up the Arduino test code and this still works (flash is still ok)

**Comment by RudolphRiedel** [comment]
I currently have no idea why flashing any of the .blob even can fail.
And right now I can not build eve_flash_pico.uf2.
But the issue I put in was a friendly knock on the frontdoor and the issue already has an assignee. :-)

For now flashing the .blob from a program should help.

In tft.c with "#define TEST_UTF8 0" copy these three lines:
EVE_cmd_inflate(0, flash, sizeof(flash)); /* de-compress flash-image to RAM_G */
datasize = EVE_cmd_getptr(); /* we unpacked to RAM_G address 0x0000, so the first address after the unpacked data also is the size */
EVE_cmd_flashupdate(0,0,4096); /* write blob first */

And put them right after "#endif /* TEST_UTF8 */".

Run once to install the .blob which actually is the bt815.blob and flashing the image should work.
python uf2conv4eve.py -i flash.bin

Or well, writing a writeflashblob.uf2 would be possible. :-)

**Comment by RudolphRiedel** [comment]
Ok, I had to implement this real quick. :-)
[pi-pico_flashblob.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11589253/pi-pico_flashblob.zip)
[EVE_FlashBlob_Arduino_PlatformIO.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11589262/EVE_FlashBlob_Arduino_PlatformIO.zip)

This only writes the 4k flash-blob.
I erased the flash.
Used the pi-pico_flashblob.uf2.
python uf2conv4eve.py -i flash.bin
Ran the example without flashing - works.

**Reply by graemecoard** [reply]
Ok, followed as best I could. 

- Clear with by uncommenting `EVE_cmd_flasherase` below, then uncomment and re-upload before flashing.
- Convert this file to ARGB4X4 in EAB. 
 ![test1](https://github.com/RudolphRiedel/FT800-FT813/assets/99002082/a15bed67-ca92-4cb5-9ab5-a2e5d67f7cca)
- Use "flash image generator" to convert the .raw converted image to .bin (which by default includes the .blob, is this a problem?)
- Upload using `python uf2conv4eve.py -i flash.bin`, shows success, bteve drive appears then goes away once written.
- Run this program below, modified from your example.
- I get a 100x100 purple square :(


```
void setup()
{
    pinMode(EVE_CS, OUTPUT);
    digitalWrite(EVE_CS, HIGH);
    pinMode(EVE_PDN, OUTPUT);
    digitalWrite(EVE_PDN, LOW);

    /* we are not using the Arduino SPI class since we want DMA */
#if defined(ESP32) || (ARDUINO_NUCLEO_F446RE) || (WIZIOPICO) || (PICOPI)
    EVE_init_spi();
#else
    SPI.begin(); /* sets up the SPI to run in Mode 0 and 8 MHz */
    SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
#endif

    if (E_OK == EVE_init()) /* make sure the init finished correctly */
    {
        //EVE_cmd_flasherase();
        EVE_cmd_inflate(0, flashblob, sizeof(flashblob)); /* de-compress flash-image to RAM_G */
        EVE_cmd_flashupdate(0, 0, 4096); /* write blob */
        if (E_OK == EVE_init_flash()) {
            EVE_cmd_flashread(0, 4096, 10048);
            Serial.println("Flash init success");
        }
    }


}

void loop()
{
    EVE_start_cmd_burst();
    EVE_cmd_dl_burst(CMD_DLSTART); /* Start the display list */
    EVE_cmd_dl_burst(DL_CLEAR_COLOR_RGB | 0x000000); /* set the default clear color to white */
    EVE_cmd_dl_burst(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG); /* clear the screen - this and the previous prevent artifacts between lists, Attributes are the color, stencil and tag buffers */
    EVE_color_rgb_burst(0xFFFFFF);
    EVE_cmd_text_burst(200, 50, 29, 0, "Test");
    EVE_cmd_setbitmap_burst(0, EVE_ASTC_4X4, 100, 100);
    EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
    EVE_cmd_dl_burst(VERTEX2F(0, 0));
    EVE_cmd_dl_burst(DL_END);
    EVE_cmd_dl_burst(DL_DISPLAY); /* instruct the co-processor to show the list */
    EVE_cmd_dl_burst(CMD_SWAP); /* make this list active */
    EVE_end_cmd_burst();
}
```

**Comment by RudolphRiedel** [comment]
I tried the same but modified the code a little:

`
 void setup()
 {
    pinMode(EVE_CS, OUTPUT);
    digitalWrite(EVE_CS, HIGH);
    pinMode(EVE_PDN, OUTPUT);
    digitalWrite(EVE_PDN, LOW);

 /* we are not using the Arduino SPI class since we want DMA */
 #if defined(ESP32) || (ARDUINO_NUCLEO_F446RE) || (WIZIOPICO) || (PICOPI)
    EVE_init_spi();
 #else
    SPI.begin(); /* sets up the SPI to run in Mode 0 and 8 MHz */
    SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
 #endif

    if(E_OK == EVE_init()) /* make sure the init finished correctly */
    {
 //        EVE_cmd_inflate(0, flashblob, sizeof(flashblob)); /* de-compress flash-image to RAM_G */
 //        EVE_cmd_flashupdate(0,0,4096); /* write blob */

        if (E_OK == EVE_init_flash())
        {
            EVE_cmd_flashread(0, 4096, 10048);
 //            Serial.println("Flash init success");
        }

    EVE_start_cmd_burst();
    EVE_cmd_dl_burst(CMD_DLSTART); /* Start the display list */
    EVE_cmd_dl_burst(DL_CLEAR_COLOR_RGB | 0x000000); /* set the default clear color to white */
    EVE_cmd_dl_burst(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG); /* clear the screen - this and the previous prevent artifacts between lists, Attributes are the color, stencil and tag buffers */
    EVE_color_rgb_burst(0xFFFFFF);
    EVE_cmd_text_burst(200, 50, 29, 0, "Test");
    EVE_cmd_setbitmap_burst(0, EVE_ASTC_4X4, 100, 100);
    EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
    EVE_cmd_dl_burst(VERTEX2F(0, 0));
    EVE_cmd_dl_burst(DL_END);
    EVE_cmd_dl_burst(DL_DISPLAY); /* instruct the co-processor to show the list */
    EVE_cmd_dl_burst(CMD_SWAP); /* make this list active */
    EVE_end_cmd_burst();
    }
 }

 void loop()
 {
 }
`

First of all, why the heck is using "add code" not working for me? :-)
Then I used the freshly modified uf2conv4eve.py from Bridgetek to flash the image.
And on the first try I did not get a purple square but also not the image either, more random white pixels in a otherwise black square.

On the second try I added the image a second time, converted to 8x8.
And I almost got the image correctly.
I forgot to disable the EDF block though and the image actually was written to 4224
After correcting this the test image is displayed now.

Hmm, what about the second image?
EVE_cmd_flashread(0x2800, 14272, 2752);

    EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
    EVE_cmd_setbitmap_burst(0, EVE_ASTC_4X4, 100, 100);
    EVE_cmd_dl_burst(VERTEX2F(0, 0));
    EVE_cmd_setbitmap_burst(0x2800, EVE_ASTC_8X8, 100, 100);
    EVE_cmd_dl_burst(VERTEX2F(0, 16*110));
    EVE_cmd_dl_burst(DL_END);

This is showing two fine test images right now.

Ok, back to the flash image with a single picture, this time without the EDF block.

Hmm, flashing does not seem to work, the image still is located at 4224.
And now EVE_cmd_flasherase() is not working - after the command the flash is still not empty.

So I used EVE_cmd_flasherase(); from the other programm and not it worked.
Right now a purple square is displayed.

Flashing is not working anymore though.
Looks like for some reason the flash is locking up.

Ugh, ok, looks like this is broken differently now.

Edit:
I finally made it flash:
python uf2conv4eve.py -p -f BT817 bt817.blob -i test_one_image.bin

Note the "-p".

And right now I have the white square with "Test" displayed, read from 4096.

Edit2:
EVE_cmd_flasherase() worked when I put in after EVE_init_flash().
So probably the flash needs a little more time after EVE_init() to make it to BASIC.

Hmm, nope, even with a delay(5000) in between the flash is not getting erased - odd.
Reading REG_FLASH_STATUS right after EVE_init() is returning "2".
So the flash is in state BASIC and CMD_FLASHERASE should work.

And now it worked without the delay() but after reading of REG_FLASH_STATUS?

Very odd.

**Reply by graemecoard** [reply]
Ok, I think I am going to go and buy a MPSSE programmer because I can just not get this to work, or just copy massive amounts of data into a program, and hopefully BRT will update their software to "just work"

The only way I seem to be able to get flash to erase and initialize is by using the code to copy the blob.

Although I can't be the only one with this problem, am I doing something wrong in EAB?
1. Convert image to ASTC4x4
2. Use flash image generator (select xxxASTC_4x4_KHR.raw file) and generate, select either bt817 or bt815, neither seems to work
3. Ensure to erase flash and make sure I get "flash init success" (which I only get by using your program to write blob)
4. Connect Pico as per pins in EAB Read Me
5. Run uf2conv4eve command line, as per above post (and just for good measure, tried other commands), hangs when BTEVE_FLASH shows up, light on pico stays on. (I can drag a converted flash.bin->flash.uf2 in here, but still the same outcome)
6.  Run program, and a purple square again.

P.S I have found for long areas of code you need to use ``` at each end

Thanks again

**Reply by RudolphRiedel** [reply]
It's getting fixed. :-)

This is from a preliminary version from https://github.com/Bridgetek/pico-brteve/issues/28 and I added one line extra to uf2conv4eve.py:
[eve_flash_pico.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11667887/eve_flash_pico.zip)

Something like this just works now:
python uf2conv4eve.py -p int-7 pwd-6 -f BT817 BT817.blob -i funtype.bin

The Pi-Pico needs to be in Boot-Mode.

**Reply by graemecoard** [reply]
Well it worked, once, but then I tried to upload a different image and now I just get a white image, no matter what I do, strange. eve_init_flash returns 0.

**Reply by RudolphRiedel** [reply]
Have you removed the manual writing of the blob from the controller?
If not, please do and perhaps use EVE_cmd_flasherase() once.

**Reply by graemecoard** [reply]
Flash has been erased and the blob writing portion removed. 

What about how EAB includes the blob when creating the flash.bin? Could this be an issue?

**Reply by RudolphRiedel** [reply]
>What about how EAB includes the blob when creating the flash.bin? Could this be an issue?

No, the Pi-Pico flashtool completely ignores the first 4k of the flash-file.

**Reply by graemecoard** [reply]
Would you be able to upload your flash file along with the code where it is working so I can try it out at my end? Thanks

**Reply by RudolphRiedel** [reply]
I did a slightly more comprehensive test.
First I got 8 free pictures or rather the thumbnails of 8 free pictures as these are 300x230 already.
I converted all 8 images to ASTC 8x8 with EAB 2.8.0.
Then I put Funtype_24_ASTC.glyph, Funtype_24_ASTC.xfont and 4 pictures in one flash image with EAB, and the font and the other four pictures in the second flash image.

As the pictures all have the same size and all use ASTC 8x8, the flash images only differ in content with the pictures at the same addresses and a fixed offset from one picture to the next.

I flashed the first flash-image to the display:
python uf2conv4eve.py -p int-7 pwd-6 -f BT817 BT817.blob -i Test_Image_1.bin

I modified my Arduino demo, set EVE_RVT50H in platformio.ini, in tft.c I set "#define TEST_UTF8 1" to use the font.
Then I added couple of lines to TFT_display(), line 476 to 489 to display a different image every two seconds.

I planned to run this code from the ESP32 but since I already had the Pi-Pico wired up and would need to swap back and forth, I just have it running on the Pi-Pico now.

I need to put the Pi-Pico into boot mode (the "RPI-RP2" drive pops up, then I can flash the image:
python uf2conv4eve.py -p int-7 pwd-6 -f BT817 BT817.blob -i Test_Image_2.bin

This changes the drive to "BTEVE-FLASH", displays "Bridgetek BT40817" and "RP2040 FlashTool" on the top left, then a matrix-style animation when flashing.
When the flashing is done the drive is changed back to "RPI-RP2" and the screen goes dark.

After a reset my application is running again and displays the other four images.

Here, this is everything, from the pictures to uf2conv4eve.py:
[Test_PiPiCo_Flash.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11726118/Test_PiPiCo_Flash.zip)

**Comment by graemecoard** [comment]
Ok, I tried that, flashing appears to work, not sure what you mean by matrix-style animation but I get a faint white gradient that sort of moves on the screen during upload. 
But when I load the program supplied I get the background, star etc but no files that are supposed to be loaded from the flash.
I loaded the demo onto the Pi just to make sure the esp32 wasn't the culprit.
I checked pin assignments on the pi, and I'm thinking if it displays the demo, the pins should be correct
Starting to look like it's hardware....
Perhaps I should contact riverdi?

Thanks again for the help.

**Comment by RudolphRiedel** [comment]
>Ok, I tried that, flashing appears to work, not sure what you mean by matrix-style animation

The screen fills with a green-ish chaos of characters.
https://user-images.githubusercontent.com/48436368/242782521-a42280db-6601-41a9-ace6-9697b6043f9a.png

>but I get a faint white gradient that sort of moves on the screen during upload.

Oh, how do you power the display?
The backlight is set rather high during flashing, if I remember correctly the setup used 2.5W during flashing, excluding the Pi-Pico.
I am using one of my adapter boards that has a switching regulator and apply 8V from a bench PSU.

>But when I load the program supplied I get the background, star etc but no files that are supposed to be loaded from the flash.

Odd, it should not even work if the font is not present.
And if the images are not there it probably displays a pink-ish rectangle.

>I checked pin assignments on the pi, and I'm thinking if it displays the demo, the pins should be correct

Yes, the pin-change in the command line is to have the same pin assignment than what I am using with the software.
And I do not have the INT pin connected at all.

**Reply by graemecoard** [reply]
Backlight is powered by a separate usb connection. Surely if the pico can display the demo it is working fine?
It's very strange that on a couple of occasions, my first attempt has almost worked and then northing works at all after that.

**Reply by RudolphRrr** [reply]
The backlight is set to a lower value when running the demo, this is quite a difference in power consumption:
EVE_memWrite8(REG_PWM_DUTY, 0x30);

You could set this to 0x60 and check if the demo is still running.

**Reply by graemecoard** [reply]
Tried at 0x80, no problems.

**Reply by RudolphRrr** [reply]
That may leave the logic supply, my adapter is providing 3.35V for both the logic and the backlight and way more current than the display could draw.
I am not sure but the display might draw more current on the logic supply during writing to the flash.
But in general, yes, if the demo works flashing should be fine as well.

**Reply by RudolphRrr** [reply]
Ok, I checked a few NOR Flash SPI datasheets and the difference between reading and writing is nothing to be concerned about.
